govcd/vm.go

```diff
@@ -850,3 +851,186 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
 
 	return vm.GetGuestCustomizationSection()
 }
+
+// AddInternalDisk creates disk type *types.DiskSettings to the VM.
+// Returns new disk ID and error.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) AddInternalDisk(diskData *types.DiskSettings) (string, error) {
+	err := vm.Refresh()
```

Comment:
We agreed within Dainius comment that's it best scenario to have enforced(implicit) refresh as we handle very important thing - internal disk. As any disk change involves passing all set of internal disk we need to work with latest info.

Reply:
Let me rephrase: I see that we don't check whether the disk we want to add exists already.
Given that, why do we need to refresh? If the disk addition operation is idempotent, all is fine and we don't need a refresh. If it isn't, then after refresh we need to check whether the disk exists. Otherwise, we are just wasting time.

@Didainius, @lvirbalas , your take?

Reply:
Just a few lines below there is a check if `diskSettings == nil` and also appending to empty diskSettings slice. If someone initialized the VM with just HREF - he could actually send an empty disk payload and wipe all the disks. At least that was how I understood.

Reply:
Normally when you add Disk, you don't provide diskId, as vCD generates it. So you want to check if disk exist by adapter type,bus number, unit number and all other details? I think we will get to may strange situations and I recommend to leave that to API. `Refresh` is needed as we send all existing disk with new one together as new request to `reconfigureVM`

Reply:
What would happen if the combination of type, bus number, and unit number is already taken? Would the current disk go to overwrite the existing one?

Reply:
If it can't create disk due bus type taken or something else, API sends back error with message. You can't overwrite disk.